libs/datastructures/include/hpx/datastructures/tuple.hpp

```diff
@@ -468,6 +471,14 @@ namespace hpx { namespace util {
     };
 #endif
 
+#if defined(HPX_HAVE_CXX17_STD_VARIANT)
+    template <typename... Ts>
+    struct tuple_size<std::variant<Ts...>>
+      : std::integral_constant<std::size_t, sizeof...(Ts)>
+    {
+    };
+#endif
```

Comment:
A variant is not a tuple like thing. Further, the active member of the variant is not a compile time property.

Reply:
And still `std::get` is defined for it... (https://en.cppreference.com/w/cpp/utility/variant/get)

Reply:
Well, `std::get` is defined for it, true. But `std::get` is not (generically) implemented in terms of `std::tuple_size`/`std::tuple_element`.

Reply:
What this additionally enables is structured binding for variants:
```
std::variant<int, float, double> v = ...;
auto [i, f, d] = v;
```
Which you have to agree, does not make a lot of sense.

Reply:
I agree that we don't need the specialization for `tuple_size`. We do however need it for `tuple_element` (for instance here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/datastructures/include/hpx/datastructures/tuple.hpp#L53). Alternatively we would have to redo all of our implementation for `util::get`.

Reply:
Separate implementations of `util::adl_barrier::get` for `std::variant` would do the trick as well.